set up the template

kalenkovich opened the pull request:
I had to start from scracth. There is still no actual frequency-tagging code added yet. Just making sure everything works, tests are passed, docs are built, etc.

I did pretty much just what is described in the docs for the mne-tools project template [here](https://mne.tools/mne-project-template/quick_start.html). Not everything worked straight away though - thus a thousand commits. Also, I am not sure what half the CI-like services are doing but I connected them anyway. Appveyor failing is fine, by the way: I haven't told it what to actually do yet.

@dominikwelke, could you please check that you can build the package, the docs and the tests are passed locally? I put the instructions under contributing.md.

*Just complaining:* The darndest thing is that it took me an extra million years just because `open` and `codecs.open` behave a bit differently on Windows. Because of that I always got an error when trying to deploy to PyPI. And the error message was so "informative" and the error itself was so common that I found this image long before I actually figured out how to fix it:

![](https://pbs.twimg.com/media/EwTtDP0XMAMPuaJ.jpg)

kalenkovich committed:
update files manually

kalenkovich committed:
installed pytest

kalenkovich committed:
fix: conda environment file

- tell conda to look for mne in the conda-forge channel

Otherwise, (by default) it only looks in the "defaults" channell which does not have mne.
I specified the channel at the package level instead of the environment level because in my expereinece it leads to more reproducible environments.
Of course, we are not specifying package versions so any actual reproducibility is out of the question.
Still, at least having most of the packages from the "defaults" channels instead of an unpredictable mix of "defaults" and "conda-forge" is already something.

- specify the Python version - 3.7

Without this, conda took forever to solve the environment.
The reason for choosing 3.7 is simply that 3.8 feels too new for me at the moment.

kalenkovich committed:
installed docs dependencies

kalenkovich committed:
fix: remove ref to deleted docs file

kalenkovich committed:
add contributing.md

kalenkovich committed:
added test coverage

- installed `pytest-cov`
- added instructions for testing the coverage to `contributing.md`
- added a test for the Bar class

kalenkovich committed:
fix: don't save a test plot

Just annoying, nothing serious.

kalenkovich committed:
deploy to PyPI

- install twine to check and  upload to PyPI
- use `open` instead of `codecs.open` to avoid messing up the line endings on Windows
- set `long_description_content_type` to avoid a warning during `twine check`

pep8speaks commented:
Hello @kalenkovich! Thanks for updating this PR. We checked the lines you've touched for [PEP 8](https://www.python.org/dev/peps/pep-0008) issues, and found:









There are currently no PEP 8 issues detected in this Pull Request. Cheers! :beers: 

##### Comment last updated at 2021-04-25 14:54:13 UTC

kalenkovich committed:
fix: do not use the original template repo anywhere

kalenkovich commented:
There are some pep8 issues - I'll fix those before merging.

dominikwelke commented:
- [x] setup works
- [x] coverage test works
- [x] tests passed
- [x] doc builds

two things though:
- i'll make some inline suggestions to the instructions in `contributing.md`. 
- creating the conda environment from `environment.yml`: 
i'm not super familiar with it, but if i read the file correctly `numpy`, `sciypy`, and `mne` are fetched twice - once via `conda` in the `environment.yml` file, and once via `pip` in the `requirements.txt` file.  i assume conda will have priority. we could eliminate them from either of the requirement files

dominikwelke commented on contributing.md:
```diff
@@ -0,0 +1,76 @@
+
+## Check test coverage
+
+From `freqtag`:
```
this is a bit confusing since all three the package, the conda environment, and the folder we cd to are called freqtag :)

dominikwelke committed:
Update contributing.md

dominikwelke commented:
> _Just complaining:_ The darndest thing is that it took me an extra million years just because `open` and `codecs.open` behave a bit differently on Windows. Because of that I always got an error when trying to deploy to PyPI. And the error message was so "informative" and the error itself was so common that I found this image long before I actually figured out how to fix it:
> 
> ![](https://camo.githubusercontent.com/cc24be22f1e3a730875f6b51397cec14f1e277ab1a8992bf66f2f1af49040cfc/68747470733a2f2f7062732e7477696d672e636f6d2f6d656469612f45775474445030584d414d5075614a2e6a7067)

:D

kalenkovich commented:
> * [x]  setup works
> * [x]  coverage test works
> * [x]  tests passed
> * [x]  doc builds

Thanks for checking!
 
> two things though:
> 
> * i'll make some inline suggestions to the instructions in `contributing.md`.

A bit confused about that one. You've already committed your suggestions, right?

> * creating the conda environment from `environment.yml`:
>   i'm not super familiar with it, but if i read the file correctly `numpy`, `sciypy`, and `mne` are fetched twice - once via `conda` in the `environment.yml` file, and once via `pip` in the `requirements.txt` file.  i assume conda will have priority. we could eliminate them from either of the requirement files

`conda` won't automatically install stuff from `requirements.txt` so there won't be any double installation. I am not entirely sure but I think some of the CIs might set up the environment using `requirements.txt` so I'd rather keep it.

dominikwelke commented:
> A bit confused about that one. You've already committed your suggestions, right?

yes, i was confused too :D 
my *intention* was to suggest changes, but then i *did* a normal commit.. 
can you remind me how to suggest changes via the github website? ppl in the mne prs do it all the time, but i didnt find it earlier.

> `conda` won't automatically install stuff from `requirements.txt` so there won't be any double installation. I am not entirely sure but I think some of the CIs might set up the environment using `requirements.txt` so I'd rather keep it.

alright, my bad.. i was reading to fast - i thought it would pip install the `requirements.txt` in root as well 
```
  - pip:
      - -r file:freqtag/tests/requirements.txt
      - -r file:doc/requirements.txt
      - twine
```

kalenkovich commented:
> yes, i was confused too :D
> my intention was to suggest changes, but then i did a normal commit..

Ah. alright. My intentions and my results often diverge as well when it comes to GitHub :) My first ever PR: just merged it myself immediately after creating it. We then decided to change some settings so I wouldn't be able to merge at all :)

> can you remind me how to suggest changes via the github website? ppl in the mne prs do it all the time, but i didnt find it earlier.

Sure! You go under the "Files changed" tab. There, you select a line or a couple of lines and press "+" to add a comment. In the comment toolbar above the editing area, there is a button with a plus-minus sign which allows you to add a suggestion alongside the comment.

kalenkovich commented:
Since otherwise everything seems to work, I'll just fix the PEP8 issues and merge.

kalenkovich committed:
install flake8 and fix some issues

- long lines
- unused imports
- imports not at the top
- no space between # and the comment text
- non-pretty indentations in lists/dicts
- `u` before strings

In doc/conf.py:80: the ''.format() expects {} not %s
